File: src/main/java/com/pandus/leetcode/solutions/google/LengthOfLongestSubarrayWithAtMostKFrequency.java
package com.pandus.leetcode.solutions.google;

import java.util.HashMap;
import java.util.Map;

public class LengthOfLongestSubarrayWithAtMostKFrequency {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int ans = 0;
        int left = 0;
        for (int right = 0; right < nums.length; right++) {
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);
            while (map.get(nums[right]) > k) {
                map.put(nums[left], map.get(nums[left]) - 1);
                left++;
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/daily/LengthOfLongestSubarrayWithAtMostKFrequency.kt
package com.pandus.leetcode.solutions.daily

import kotlin.math.max

// Reference:https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency
class LengthOfLongestSubarrayWithAtMostKFrequency {
    fun maxSubarrayLength(nums: IntArray, k: Int): Int {
        var ans = 0
        var start = -1
        val frequency = LinkedHashMap<Int, Int>()
        for (idx in nums.indices) {
            frequency[nums[idx]] = frequency.getOrDefault(nums[idx], 0) + 1
            while (frequency[nums[idx]]!! > k) {
                start++
                frequency[nums[start]] = frequency[nums[start]]!! - 1
            }
            ans = max(ans, (idx - start))
        }
        return ans
    }
}
